File: Core/src/services/subdomain_service.py
import grpc

from configs.redis_config import RedisConfig
from controllers.subdomain_controller import SubdomainController
from proto.v1.subdomain.model_pb2 import Subdomain
from proto.v1.subdomain.service_pb2 import GetOccupiedSubdomainsRequest, GetOccupiedSubdomainsResponse
from proto.v1.subdomain.service_pb2_grpc import SubdomainServiceServicer


class SubdomainService(SubdomainServiceServicer):
    async def GetOccupiedSubdomains(
            self,
            request: GetOccupiedSubdomainsRequest,
            context: grpc.aio.ServicerContext) -> GetOccupiedSubdomainsResponse:
        async with RedisConfig() as session:
            subdomains = [
                Subdomain(name=subdomain_name)
                for subdomain_name in await SubdomainController(session).get_all_subdomains()
            ]
        
        return GetOccupiedSubdomainsResponse(subdomains=subdomains)

File: core/internal/grpc/subdomain_service/subdomain_service.go
package subdomain_service

import (
	"context"

	proto "github.com/mekstack/nataas/core/proto/pb"

	"google.golang.org/grpc"
)

type SubdomainController interface {
	GetOccupiedSubdomains(context.Context) ([]*proto.Subdomain, error)
}

type service struct {
	proto.UnimplementedSubdomainServiceServer
	subdomainController SubdomainController
}

func Register(server *grpc.Server, controller SubdomainController) {
	proto.RegisterSubdomainServiceServer(server, &service{
		subdomainController: controller,
	})
}

func (s *service) GetOccupiedSubdomains(ctx context.Context, request *proto.GetOccupiedSubdomainsRequest) (*proto.GetOccupiedSubdomainsResponse, error) {
	pool, err := s.subdomainController.GetOccupiedSubdomains(ctx)

	if err != nil {
		return nil, err
	}

	return &proto.GetOccupiedSubdomainsResponse{
		Subdomains: pool,
	}, nil
}
